tests/golden/openshift-rbac-proxy/thanos/thanos/query/deployment.yaml

```diff
@@ -0,0 +1,97 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app.kubernetes.io/component: query-layer
+    app.kubernetes.io/instance: thanos-query
+    app.kubernetes.io/name: thanos-query
+    app.kubernetes.io/version: v0.18.0
+  name: thanos-query
+  namespace: syn-thanos
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app.kubernetes.io/component: query-layer
+      app.kubernetes.io/instance: thanos-query
+      app.kubernetes.io/name: thanos-query
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/component: query-layer
+        app.kubernetes.io/instance: thanos-query
+        app.kubernetes.io/name: thanos-query
+        app.kubernetes.io/version: v0.18.0
+    spec:
+      affinity:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+            - podAffinityTerm:
+                labelSelector:
+                  matchExpressions:
+                    - key: app.kubernetes.io/name
+                      operator: In
+                      values:
+                        - thanos-query
+                namespaces:
+                  - syn-thanos
+                topologyKey: kubernetes.io/hostname
+              weight: 100
+      containers:
+        - args:
+            - query
+            - --grpc-address=0.0.0.0:10901
+            - --http-address=0.0.0.0:9090
+            - --log.level=info
+            - --log.format=logfmt
+            - --query.replica-label=prometheus_replica
+            - --query.replica-label=rule_replica
+            - --query.auto-downsampling
+          env:
+            - name: HOST_IP_ADDRESS
+              valueFrom:
+                fieldRef:
+                  fieldPath: status.hostIP
+          image: quay.io/thanos/thanos:v0.18.0
+          livenessProbe:
+            failureThreshold: 4
+            httpGet:
+              path: /-/healthy
+              port: 9090
+              scheme: HTTP
+            periodSeconds: 30
+          name: thanos-query
+          ports:
+            - containerPort: 10901
+              name: grpc
+            - containerPort: 9090
+              name: http
+          readinessProbe:
+            failureThreshold: 20
+            httpGet:
+              path: /-/ready
+              port: 9090
+              scheme: HTTP
+            periodSeconds: 5
+          resources: {}
+          terminationMessagePolicy: FallbackToLogsOnError
+        - args:
+            - --upstream=http://0.0.0.0:9090
```

Comment:
I'm really surprised this works. Shouldn't it be `127.0.0.1`?

Reply:
not really. The query container listens on 0.0.0.0 as you can inspect in the golden files:
```yaml
      containers:
        - args:
            - query
            - --grpc-address=0.0.0.0:10901
            - --http-address=0.0.0.0:9090
```

Reply:
I always thought `0.0.0.0` is not a valid destination address.

Apparently while technically not valid `0.0.0.0` may work in practice https://superuser.com/questions/949428/whats-the-difference-between-127-0-0-1-and-0-0-0-0 .

Reply:
Well, to quote your own article:

> In the context of servers, 0.0.0.0 means "all IPv4 addresses on the local machine"

If query would only bind loopback with `query - --http-address=127.0.0.1:9090` then it wouldn't be reachable from outside the container.

Reply:
Yes I'm not only talking about `--upstream=http://0.0.0.0:9090` since that's the client. The listener is configured with `--http-address=http://:8080` which tells go to listen on `0.0.0.0` on both IPv4 and IPv6.